docs/content/en/docs/Design/do_transformation.md

```diff
@@ -0,0 +1,179 @@
+```mlir
+ %sum_0 = arith.constant 0.0 : f32
+ // The for-loop's bound is a fixed constant
+ %sum = affine.for %i = 0 to 10 step 2
+ iter_args(%sum_iter = %sum_0) -> (f32) {
+   %t = affine.load %buffer[%i] : memref<1024xf32>
+   %sum_next = arith.addf %sum_iter, %input : f32
+   affine.yield %sum_next : f32
+ }
+ ...
+```
+- `scf.for`: Similar to affine.for operation, scf.for operation requires a body region that exactly contains one block that [terminates with `scf.yield`](https://mlir.llvm.org/docs/Dialects/SCFDialect/#scffor-scfforop), which makes implementing early loop exits within the same level of abstraction a challenge. However, scf.for does allow input-dependent conditionals which does not adhere to data-obliviousness constraints. A solution to this could be to either have the programmer or the compiler specify an input-independent upper bound so we can transform the loop to use this upper bound and also carefully update values returned from the for-loop using conditional predication. Our current solution to this is for the programmer to add the lower bound and worst case upper bound in the static affine loop's `attributes` list.
+```mlir
+func.func @my_function(%value: i32 {secret.secret}, %inputIndex: index {secret.secret}) -> i32 {
+  ...
+  // Violation: for-loop uses %inputIndex as upper bound which causes a secret-dependent control-flow
+  %result = scf.for %iv = %begin to %inputIndex step %step_value iter_args(%arg1 = %value) -> i32 {
+    %output = arith.muli %arg1, %arg1 : i32
+    scf.yield %output : i32
+  }{lower = 0, upper = 32}
+  ...
+}
+
+// After applying Loop-Transformation
+func.func @my_function(%value: i32 {secret.secret}, %inputIndex: index {secret.secret}) -> i32 {
+  ...
+  // Build for-loop using lower and upper values from the `attributes` list
+  %result = affine.for %iv = 0 to  step 32 iter_args(%arg1 = %value) -> i32 {
+    %output = arith.muli %arg1, %agr1 : i32
+    %cond = arith.cmpi eq, %iv, %inputIndex : index
+    %newOutput = arith.select %cond, %output, %arg1
+    scf.yield %newOutput : i32
+    }
+  ...
+}
+```
+- `scf.while`: This operation represents a generic while/do-while loop that keeps iterating as long as a condition is met. An input-dependent while condition introduces a data-dependent control flow that violates data-oblivious constraints. It's challenging to implement early-exits since the scf.while is also expected to [terminate with `scf.yield`](https://mlir.llvm.org/docs/Dialects/SCFDialect/#scfwhile-scfwhileop). We are working on adding early-exit support to high-level MLIR as described in this [issue](https://github.com/google/heir/issues/922). For this transformation, the programmer needs to add the `max_iter` attribute that describes the maximum number of iterations the loop runs which we then use the value to build our static `affine.for` loop.
```

Comment:
> We are working on adding early-exit support

Generally speaking it's best to leave "we will do" statements out of documentation. They often do not get completed (for better or worse) and it just becomes documentation rot. Or they get completed and the doc becomes stale as well. There's no shame in it, it's just better to state what the code does now as well as possible. Updating the docs to reflect the future will be a job for the people who implement those features.

Best would be to add a callout block like this near the start of the section

```
> [!NOTE]  
> Early exiting from loops is not supported in `scf` and `affine`, 
> so early exits are not supported in this pipeline.
> TODO(#922): support early exits
```

Then the TODO would be linked to this docs so people know to update it.